### DISCUSSION TITLE
Best way to reuse mocks across several tests?

### DISCUSSION THREAD

**Discussion by ghost** [MikeSchulze/gdUnit3#389] [FAQ] [answered]
Sorry in advance if I'm being dumb, I'm rather new to GdUnit3 and unit testing in general. I searched in the docs for a clue on this topic as well, but couldn't find anything.

I have a couple of tests on a class, and I am setting up mocks for each test. There are common mocks between the two, so I've combined all my mocking logic into a separate function, and just call that function from within the tests themselves.

When I'm running the tests as shown in the attached [test_modal_results.gd.txt](https://github.com/MikeSchulze/gdUnit3/files/12168991/test_modal_results.gd.txt), it takes around 450ms to run each one, due to having to call the function and do the work of creating the mocks separately in each test.

### My (misguided?) attempts at optimizing test performance

As an experiment, I tried a couple of things:

1. **Creating `mocks` as a class property, and populating it during _ready()**
In this case, the tests run successfully and much faster (16-17ms) in normal mode, but in debug mode I get the following error:

```
E 0:00:01.932   get_meta: The object does not have any 'meta' values with the key 'MEMORY_POOL'.
  <C++ Error>   Method failed. Returning: Variant()
  <C++ Source>  core/object.cpp:1069 @ get_meta()
  <Stack Trace> GdUnitMockBuilder.gd:162 @ build()
                GdUnitTestSuite.gd:156 @ mock()
                modal_results_test.gd:18 @ setup_mocks()
                modal_results_test.gd:14 @ _ready()
                GdUnitRunner.gd:76 @ _process()
```

2. **Creating `mocks` as a class property, populating it during the first test, and checking for the property during the second test in order to reuse it.**
In this case, again the tests run successfully in normal mode. The first test is ~450ms while the second test is only ~4ms (might be misleading in this case, given the errors). In debug mode, I get the following errors:

```
E 0:00:02.635   get_named: Attempted get on a deleted object.
  <C++ Source>  core/variant_op.cpp:1693 @ get_named()
  <Stack Trace> modal_results_test.gd:73 @ test_calculate_levels()
                _TestCase.gd:50 @ execute()
                GdUnitExecutor.gd:217 @ execute_test_case_single()
                GdUnitExecutor.gd:348 @ Execute()
                GdUnitExecutor.gd:44948112 @ execute_test_case_single()
                GdUnitExecutor.gd:44949392 @ test_after()

E 0:00:02.635   has_method: Attempted method check on a deleted object.
  <C++ Source>  core/variant_call.cpp:1434 @ has_method()
  <Stack Trace> modal_results_test.gd:73 @ test_calculate_levels()
                _TestCase.gd:50 @ execute()
                GdUnitExecutor.gd:217 @ execute_test_case_single()
                GdUnitExecutor.gd:348 @ Execute()
                GdUnitExecutor.gd:44948112 @ execute_test_case_single()
                GdUnitExecutor.gd:44949392 @ test_after()
```

So it's pretty clear that I'm trying to do things I'm not really supposed to be doing, by structuring the tests in this way. However, I'd really like to cut down on that test time if possible, and I guess I'm just naturally inclined to try and reduce overhead and reuse data and references as much as possible.

### So my question is:

Is there a preferred, recommended way of sharing and reusing mocks across multiple tests (within the same TestSuite of course) in order to improve test performance?

Thanks for reading, and apologies for any headaches my shenanigans may have caused :P

**Comment by MikeSchulze** [comment, marked as answer]
Hello @paintedsky 

You should first read how a test-suite is organized => [TestSuite](https://mikeschulze.github.io/gdUnit3/faq/test-suite/)
What you need is a so called setup hook for a test suite.

There are two of them, `before` and `before_test`.

For example, to create a mock that applies to all tests, you should use `before`, since these objects exist over the lifetime of an testsuite.
On the other hand, `before_test`, as the name suggests, is always called before each test and is thus freshly created and only available for the lifetime of a test case.

It is important when you create a mock in `before` to reset it for each test to reset the reset the mock counted interactions.
Typically it looks like this.

```gdscript
class_name MyTest
extends GdUnitTestSuite


class MyClass:
	
	func foo() -> String:
		return "foo"


var _mock :MyClass

func before():
	_mock = mock(MyClass)
	do_return("xxx").on(_mock).foo()


func before_test():
	reset(_mock)


func test_foo():
	verify(_mock, 0).foo()
	_mock.foo()
	verify(_mock, 1).foo()


func test_foo1():
	verify(_mock, 0).foo()
```

**Reply by ghost** [reply]
Thanks, that's perfect! I'm not sure how I missed that when I was looking in the docs, but I knew there had to be a proper way to do it. Thanks again for the kind and thorough answer -- cheers! :)
